src/Mappers/BaseMapper.php

```diff
@@ -62,17 +64,158 @@ protected static function getValueFromTag(\DOMDocument $document, string $tag):
+    protected static function getField(\DOMElement $element, string $fieldTagName, $object = null): ?string
     {
         $fieldElement = $element->getElementsByTagName($fieldTagName)->item(0);
+
         if (!isset($fieldElement)) {
             return null;
         }
 
+        if (isset($object)) {
+            self::checkForMessage($object, $fieldElement);
+        }
+
         if ($fieldElement->textContent === "") {
             return null;
         }
 
         return $fieldElement->textContent;
     }
+    
+    protected static function checkForMessage($object, \DOMElement $element): void
+    {
+        if ($element->hasAttribute('msg')) {
+            $message = new Message();
+            $message->setType($element->getAttribute('msgtype'));
+            $message->setMessage($element->getAttribute('msg'));
+            $message->setField($element->nodeName);
+
+            $object->addMessage($message);
+        }
+    }
+
+    protected static function getAttribute(\DOMElement $element, string $fieldTagName, string $attributeName): ?string
+    {
+        $fieldElement = $element->getElementsByTagName($fieldTagName)->item(0);
+
+        if (!isset($fieldElement)) {
+            return null;
+        }
+
+        if ($fieldElement->getAttribute($attributeName) === "") {
+            return null;
+        }
+
+        return $fieldElement->getAttribute($attributeName);
+    }
+
+    protected static function parseBooleanAttribute(?string $value): ?bool
+    {
+        return filter_var($value, FILTER_VALIDATE_BOOLEAN);
+    }
+
+    protected static function parseDateAttribute(?string $value): ?\DateTimeImmutable
+    {
+        if ((bool)strtotime($value)) {
+            return Util::parseDate($value);
+        } else {
+            return null;
+        }
+    }
+
+    protected static function parseDateTimeAttribute(?string $value): ?\DateTimeImmutable
+    {
+        if ((bool)strtotime($value)) {
+            return Util::parseDateTime($value);
+        } else {
+            return null;
+        }
+    }
+
+    protected static function parseEnumAttribute(string $enumName, ?string $value)
+    {
+        if ($value === null) {
+            return null;
+        }
+
+        $enum = "\\PhpTwinfield\\Enums\\" . $enumName;
+
+        try {
+            $classReflex = new \ReflectionClass($enum);
+            $classConstants = $classReflex->getConstants();
+
+            foreach ($classConstants as $classConstant) {
+                if ($value == $classConstant) {
+                    return new $enum($value);
+                }
+            }
+        } catch (\ReflectionException $e) {
+            return null;
+        }
+
+        return null;
+    }
+
+    protected static function parseMoneyAttribute(?float $value): ?Money
+    {
+        if ($value === null) {
+            return null;
+        }
+
+        return Util::parseMoney($value, new Currency('EUR'));
```

Comment:
This is not good, we use Twinfield in a many currencies environment and a special treatment of EUR leads to bugs.

Please allow to currency to be passed as an argument, or remove this method.

Reply:
This is no specials treatment op EUR. There are no value fields in Twinfield directly linked to a currency, so whichever currency is used here has zero effect on functionality.

Money for PHP just needs some currency to construct an object. Could be any ISO currency.

Reply:
I'm not sure if I understand this correctly - Twinfield does not pass a currency for some amounts?

Reply:
Twinfield does not pass currency for **any** amount

Money for PHP is used when a field type in the API Documentation equals money. For example in Articles (from the API documentation):

Name | Type | Description
-- | -- | --
unitspriceexcl | money | Price excluding VAT

If you look at the XML the Money Type equals just a decimal number in Twinfield:

`<unitspriceexcl>750.00</unitspriceexcl>`

It does not include or involve a Currency.

The most basic usage in the library would be to just cast these decimals als float instead of using Money for PHP, but this would mean you don't get access to the added functionality (such as add, subtract, formater etc.). Money for PHP needs some currency code to construct so any currency code can be used, I used EUR because it was already used this way in `Transactions.` Currency operations within the money object should not be performed.

Reply:
Isn't it in the header part of the XML document? https://c3.twinfield.com/webservices/documentation/#/ApiReference/SalesInvoices

Reply:
A currency is sometimes part of the header or returned elsewhere in an entity (but more often not), but is not directly linked to individual amounts (in an attribute field on the field for example).

Reply:
Another thing to realise is that Twinfield Currency codes don't have to be ISO Currency codes, they can be anything you like. If you try to set a non existing ISO currency in Money you will get an error.

I'm also ok with just removing Money for PHP entirely and treating money fields in Twinfield as floats. Is probably the cleanest solution.

Reply:
@willemstuursma can you comment on this? My preference would be to remove Money for PHP entirely. 

After deciding on this, I think we might be ready to continue with the next bit of #142.

Reply:
Hi @iranl. We want to keep MoneyPHP\Money for sure. 

I think you should delete this method and rely on the method in Util, and always pass a currency. But it depends on the situation where it comes from.

Reply:
Did some more testing and MoneyPHP\Money does in fact accept non ISO Currencies as input. So that's one (non-existent) problem out of the way.

This function in BaseMapper does have a use though as the function in Util does not accept null input for value or currency (which are possibilities when reading from Twinfield). This functions filters those out (and returns null) and passes all other input to the function in Util.

So I believe the function should stay and I have modified it to take a currency string as input instead of hardcoding it to EUR. 

There are some problems along the way with currencies though (such as values in the reporting currency instead of the base currency). But those have solutions in other places.

Reply:
Added a function to BaseMapper to retrieve the Base and Reporting currency from the selected office for use in the specific mappers when working with currencies that are not linked to the currency in the specific object or when there is no currency field in the object